neuralpp/symbolic/expression.py

```diff
@@ -1,10 +1,43 @@
 from __future__ import annotations  # to support forward reference for recursive type reference
 
+import fractions
 import typing
+import operator
 from typing import List, Any, Optional, Type, Callable
 from abc import ABC, abstractmethod
 
-
+# typing.Callable can be ambiguous.
+# Consider the following two tests:
+#   >>> isinstance(int, Type)
+#   True
+#   >>> isinstance(1, Type)
+#   False
+# So "int is an instance of Type" and "1 is not an instance of Type"
+# But for Callable, we have
+#   >>> isinstance(Callable[[int],int], Callable)
+#   True
+#   >>> isinstance(lambda x: x, Callable)
+#   True
+# Here the `Callable` in each case means differently, which makes `Callable` ambiguous.
+#
+# Imagine a typed language, we'd have
+#   >>> isinstance(lambda (x: int) : int := x, int->int)
+#   True
+# and only one of the following two cases can be true:
+# CASE ONE: `Callable` means "the type of all function types".
+#   >>> isinstance(int->int, Callable)
+#   True
+#   >>> isinstance(lambda (x: int) : int := x, Callable)
+#   False
+# or
+# CASE TWO: `Callable` means "the type of all functions"
+#   >>> isinstance(int->int, Callable)
+#   False
+#   >>> isinstance(lambda (x: int) : int := x, Callable)
+#   True
+#
+# It should be noted that our usage of `Callable` here means the first, i.e., "the type of all function types"
+# And in our code we use `isinstance()` for cases like `isinstance(Callable[[int],int], Callable)`.
 ExpressionType = Callable | Type
```

Comment:
I got confused by return type `Callable` even though I just wrote it several days ago...So wrote these docs.